### PR TITLE
Trace requests with opentracing and jaeger

### PR DESCRIPTION
**Pull request by haletsky**

### CODE REVIEW

**Comment by msadovyi** [comment]
update mw tests, please

**Review by germankrause — APPROVED** [review]

**Review by germankrause — APPROVED** [review]

File: packages/types/src/gateway-schema.json [hunk 1]
```diff
@@ -149,6 +149,23 @@
         }
       ],
       "description": "This value specifies an HTTP Request size limit for a particular API Definition. Accepts numeric (in bytes) or string values, i.e. 'b' for bytes, 'kb' for kilobytes and 'mb' for megabytes (f.e., '10kb' or '10mb')."
+    },
+    "tracing": {
```
Comment (valcker) [inline packages/types/src/gateway-schema.json:153]: Is tracing always on? It should be possible to enable/disable it.

File: packages/types/src/gateway-schema.json [hunk 2]
```diff
@@ -149,6 +149,23 @@
...
+      "type": "object",
+      "description": "Configure Jaeger agent endpoint",
+      "additionalProperties": false,
+      "properties": {
```
Comment (valcker) [inline packages/types/src/gateway-schema.json:157]: Today we support only Jaeger but in future, we might also support Zipkin. Let's add a "tracer name" property or smth like that.

File: packages/gateway/src/middleware/utils.ts [hunk 3]
```diff
@@ -1,4 +1,5 @@
 import { Request } from 'express';
+import { Span } from 'opentracing';
```
Comment (valcker) [inline packages/gateway/src/middleware/utils.ts:2]: We import opentracing but it is included as a dev dependency in package.json. Is it normal?

File: packages/gateway/src/middleware/assign-request-id.ts [hunk 4]
```diff
@@ -11,13 +11,23 @@ const assignRequestId: RequestHandler = (req, res, next) => {
...
     query: req.body || '',
     userAgent: req.headers?.['user-agent'] || '',
     ip: req.ip,
     request: req,
     date: Date.now(),
+  };
+
+  metricEmitter.emit(MetricsChannels.GOT_REQUEST, req.id, metricData);
+
+  req.context.tracerSpan.setTag('requestId', req.id);
```
Comment (valcker) [inline packages/gateway/src/middleware/assign-request-id.ts:25]: I guess its going to do nothing if tracing is not enabled?

**Review by valcker — CHANGES_REQUESTED** [review]

**Review by valcker — APPROVED** [review]

**Pull request merged**